core/impl/src/main/java/com/blazebit/persistence/impl/AbstractCTECriteriaBuilder.java

```diff
@@ -221,18 +224,40 @@ public CTEInfo createCTEInfo() {
                 if (selectExpression instanceof NullExpression) {
                     // When binding null, we don't have to adapt anything
                 } else if (selectExpression instanceof PathExpression) {
-                    PathExpression pathExpression = (PathExpression) selectExpression;
-                    // Only append the id if it's not already there
-                    if (!idAttribute.getName().equals(pathExpression.getExpressions().get(pathExpression.getExpressions().size() - 1).toString())) {
-                        pathExpression.getExpressions().add(new PropertyExpression(idAttribute.getName()));
+                    boolean firstBinding = true;
+                    PathExpression baseExpression = ((PathExpression) selectExpression).clone(false);
+                    for (String idOrUniqueKeyProp : idOrUniqueKeyProps) {
+                        // For the first id prop, alter the existing expression, for consecutive props
+                        // add additional selects
+                        PathExpression pathExpression = firstBinding ?
+                                ((PathExpression) selectExpression) : baseExpression.clone(false);
+
+                        if (!idOrUniqueKeyProp.equals(pathExpression.getExpressions().get(pathExpression.getExpressions().size() - 1).toString())) {
```

Comment:
Can you explain why this check is necessary? IMO this is always false.

Reply:
This is necessary for two things:
1) People that bind `association.id` directly
1) Recursive calls from the same method

Reply:
Maybe I'm not understanding the code properly, but to me this seems like allowing to bind `someAssociation.sameNamedProperty` to an attribute like `id` which contains a property `sameNamedProperty`. Am I understanding this correctly?

Reply:
It seems this happend due to the code that was used before which tried to handle a second call to prepareAndGetAttributes properly by skipping the add of the id property if it was already there.

Reply:
> Maybe I'm not understanding the code properly, but to me this seems like allowing to bind someAssociation.sameNamedProperty to an attribute like id which contains a property sameNamedProperty.

Eventually, it only matters that the columns of the CTE are mapped properly. Projecting entities in CTE queries will lead to additional columns in the selection that are not defined by the CTE and potentially even messes up attribute ordering. This fix traverses the `idOrUniqueKey` attributes so that every column is mapped to the underlying referenced attribute. Ie. `someAssocitionMappedByCompoundKey` will render `someAssocitionMappedByCompoundKey.keyA, someAssocitionMappedByCompoundKey.keyB` and adjust the mapping so that each is mapped to their respective column in the cte.

The code still does exactly the same if there is only one id attributre: it adds the id attribute to the path expression if its not already there. The latter can still happen even with the added changes.

 I removed the check that was already there before and saw some tests failing, I'll get back to you as to why this happens because it seems unlogical.

Reply:
Actually, I think it's ok like this. Just re-add the comment that you removed.

Reply:
Allright so the idea behind this rewrite is that we simply can't bind association types to plain objects, and have to bind to the underlying identifier keys instead. It currently *is* however allowed, to bind an CTE association attribute directly to its identifier, i.e. `bind("someAttribute").to("someAssociation.id")` as opposed to `bind("someAttribute").to("someAssociation.id")`. There are tests that verify that both are currently allowed.

When in this method, we obtain the *identifier or unique properties* of the association attribute *on the CTE*, i.e. the inverse mapping from the join columns of the attribute. In the case a single attribute is used (most of the time) it is still possible that the binding goes to either `"someAssociation"` or `"someAssociation.id"`. In the case of the latter, we have to check whether we're not adding the property name twice.

For compound keys, it is unlikely that a user would attempt to bind to the id attribute directly, as this would involve two different bindings. I.e. there is no point in binding `bind("someAttribute").to("someAssociation.key1")`, so one would bind the keys seperately any way (`bind("someAttribute.key1").to("someAssociation.key1")). Because `"someAttribute.key1"` is not a *Joinable* attribute, the code won't even attempt to optimise this code (correctly).

So the code works, and the check has to remain in place. The only problem I have is that it is technically possible to have a collision between the association attribute name and the referenced identifier or unique key names. I.e. a case mapping `someAssoc.someId.someId` would be required, this would incorrectly be translated into `someAssoc.someId`. In order to fix this, we have to resolve the `Attribute` of the `PathExpression`, (is this possible through the `PathTargetResolvingExpressionVisitor`?), and verify that this `Attribute` is  not equal to the identifier or unique key `Attribute` on the referenced entity type.

Reply:
IMO binding `someAssociation` to `someAssocation.id` shouldn't be allowed. The thing is, that the code alters the binding and select expressions i.e. it adds the id property to the path expression. To prevent adding the id multiple times in case the `prepare` is executed multiple times, it is checked whether the path already ends with that and if so, skip adding it again. Ideally, we would split embeddables and assocations to binds of the individual properties, but I'm not sure how this could work out. I'd be very much interested in such an approach though as I was in need for something like that recently for #509.

Splitting off to individual property binds might bring a few possible problems. For once, you might have to transform a parameter by extracting an attribute value. This shouldn't be too hard, the ParameterTransformer is there to help you.
The other possible issue that comes to my mind is that you might have to order the binds in a specific way so that values are bound to the correct CTE attributes in the SQL, but I haven't given that possible problem enough thought yet. Maybe try to add a test that does a fetch join on a multi-column assocation of a CTE in the main query to verify it works properly.

Reply:
> I can alter the code to do this, it's not going to be easy though, but doing so would allow for some additional checks that might be worthwhile. I would need to resolve the attribute of the path expression against the meta model though. Is there an expression visitor for that? One issue I see is that I have to pick a class, which will mean I have to figure out whether to resolve to the query root or a join alias in the alias manager. Is such code already out there somewhere?

The PathTargetResolvingExpressionVisitor should be capable of doing that. Note that you only need to analyze this further when the bind attribute is an assocation. If it is an assocation, the expression can only be a PathExpression, so it should be easy to check.

> Do you mean to split of these individual properties at bind time? Or do you think that a similar approach, namely post processing them in prepareAndGetAttributes, would suffice? why would or wouldn't that work? As far as I can see, it should be possible to solve this similarly to how the association type is handled now.

I would split off the properties to individual binds during after the type check of the target expression.